src/wix/WixToolset.Core/ExtensibilityServices/BackendHelper.cs

```diff
@@ -119,5 +119,40 @@ public bool IsValidShortFilename(string filename, bool allowWildcards)
         {
             return Common.IsValidShortFilename(filename, allowWildcards);
         }
+
+        public bool TryParseFourPartVersion(string version, out string parsedVersion)
+        {
+            if (WixVersion.TryParse(version, out var wixVersion) && wixVersion.HasMajor && wixVersion.Major < 65536 && wixVersion.Minor < 65536 && wixVersion.Patch < 65536 && wixVersion.Revision < 65536)
+            {
+                parsedVersion = $"{wixVersion.Major}.{wixVersion.Minor}.{wixVersion.Patch}.{wixVersion.Revision}";
+                return true;
+            }
+
+            parsedVersion = null;
+            return false;
+        }
+
+        public bool TryParseMsiProductVersion(string version, bool strict, out string parsedVersion)
+        {
+            if (WixVersion.TryParse(version, out var wixVersion) && wixVersion.HasMajor && wixVersion.Major < 256 && wixVersion.Minor < 256 && wixVersion.Patch < 65536 && wixVersion.Labels == null && String.IsNullOrEmpty(wixVersion.Metadata))
+            {
+                parsedVersion = $"{wixVersion.Major}.{wixVersion.Minor}";
+
+                if (strict || wixVersion.HasPatch)
+                {
+                    parsedVersion += $".{wixVersion.Patch}";
+                }
+
+                if (!strict && wixVersion.HasRevision)
+                {
+                    parsedVersion += $".{wixVersion.Revision}";
+                }
```

Comment:
Am I reading this correctly? An MSI with `ProductVersion` v1.2.3.4 with default `MajorUpgrade` will get 1.2.3 in the upgrade table, which will make Burn think they're not the same.

Reply:
You are reading that correctly. The Upgrade table versions only support 3-part versions, the fourth part is ignored so adding that part of the version is conceptually misleading. We could add additional data in the fourth field if it helps Burn implemention additional versioning features but there shouldn't be exact comparisons between `ProductVersion` and Upgrade table rows.

Reply:
Burn v3 and v4 considers the fourth field as significant when figuring out whether the package is a downgrade or an upgrade. We have discussed changing Burn to only pay attention to the first three fields like Windows Installer but it didn't seem like anyone wanted that.

Reply:
It considers the fourth field in both the ProductVersion and the Upgrade table rows.

Reply:
Great. Then those versions should be WixVersion in Burn. I'll allow the full range ProductVersion in the Upgrade table so Burn can take advantage of it. Depending on the decision to #6788 then Burn may have lots of information to process them.